### ISSUE TITLE
Serialize based on field type

### ISSUE THREAD

**Issue by mklaber** [tkutcher/quickbase-client#47]
I'd consider submitting this as a pull request but not sure where you'd prefer it go (if anywhere). But, here are a couple changes I've made to the generated class files that make the `add_records` method actually work for us. At the very least, they probably belong in the [`QuickBaseTable`](https://github.com/tkutcher/quickbase-client/blob/dev/src/quickbase_client/orm/table.py#L51) base class.

What the following excerpts demonstrate are ways to handle:

* QB errors when sending a datetime to a date field — what's particularly annoying about the REST API is that the XML API has no problem receiving a datetime for a date field
* QB's lack of support for unicode characters — this isn't an API problem so much as it is a QB problem but since we're programatically adding data to QB, might as well nip the problem where we can

The API (and the front end) will store `Tobias Fünke` or `Lindsay Bluth Fünke` like so:
![image](https://user-images.githubusercontent.com/939816/123946955-af57fb80-d997-11eb-954c-a718d65c13e0.png)

The code I've added normalises ü to u. Note that I've explicitly decided not to include Rich Text fields in this operation because it's reasonable to assume whatever WYSIWYG that is feeding HTML to this field would URL encode these properly.

## Class Excerpts

Ideally this would go in [`QuickBaseJsonEncoder`](https://github.com/tkutcher/quickbase-client/blob/dev/src/quickbase_client/orm/serialize.py#L14-L21) but that class does not have access to the field/attribute information. So I'm currently doing the modification at the QBT setter level.

```python
class UnitApplication(QuickBaseTable):
    __dbid__ = "abcd12345"
    __app__ = MyHeyloApp

    def __setattr__(self, name: str, value: Any) -> None:
        if isinstance(value, datetime):
            fi = self.__class__.get_field_info(name)
            if fi.field_type == Qb.DATE:
                value = value.date()
        if isinstance(value, str):
            fi = self.__class__.get_field_info(name)
            if fi.field_type in [
                Qb.TEXT,
                Qb.TEXT_MULTILINE,
                Qb.TEXT_MULTI_SELECT,
                Qb.TEXT_MULTIPLE_CHOICE,
            ]:
                # from: https://stackoverflow.com/a/517974/612166
                nfkd_form = normalize("NFKD", value)
                value = "".join([c for c in nfkd_form if not combining(c)])
        return super().__setattr__(name, value)

    date_created = QuickBaseField(fid=1, field_type=Qb.DATETIME)
    date_modified = QuickBaseField(fid=2, field_type=Qb.DATETIME)
    recordid = QuickBaseField(fid=3, field_type=Qb.NUMERIC)
    record_owner = QuickBaseField(fid=4, field_type=Qb.USER)
    last_modified = QuickBaseField(fid=5, field_type=Qb.USER)

    related_unit = QuickBaseField(fid=6, field_type=Qb.NUMERIC)
    unit_name = QuickBaseField(fid=7, field_type=Qb.TEXT)
    # etc. etc. etc.
```

## Tests

Using pytest and the lovely [assertpy](https://github.com/assertpy/assertpy) library and the mediocre [Faker](https://faker.readthedocs.io/en/master/providers/baseprovider.html) library.

```python
class UnitApplicationForTesting(UnitApplication):
    date_time_as_date = QuickBaseField(fid=900, field_type=Qb.DATE)
    date_time_as_date_time = QuickBaseField(fid=901, field_type=Qb.DATETIME)
    rich_text_as_rich_text = QuickBaseField(fid=902, field_type=Qb.RICH_TEXT)
    text_as_ascii_text = QuickBaseField(fid=903, field_type=Qb.TEXT)
    text_multiline_as_ascii_text = QuickBaseField(fid=904, field_type=Qb.TEXT_MULTILINE)
    text_multi_select_as_ascii_text = QuickBaseField(
        fid=905, field_type=Qb.TEXT_MULTI_SELECT
    )
    text_multiple_choice_as_ascii_text = QuickBaseField(
        fid=906, field_type=Qb.TEXT_MULTIPLE_CHOICE
    )


test_ua_schema: UnitApplicationForTesting = UnitApplicationForTesting.schema


def test_stores_datetime_as_date() -> None:
    # arrange
    dt = fake.date_time()
    # act
    sut = UnitApplicationForTesting(date_time_as_date=dt)
    # assert
    assert_that(sut.date_time_as_date).is_equal_to(dt.date())


def test_stores_datetime_as_datetime() -> None:
    # arrange
    dt = fake.date_time()
    # act
    sut = UnitApplicationForTesting(date_time_as_date_time=dt)
    # assert
    assert_that(sut.date_time_as_date_time).is_equal_to(dt)


def test_stores_rich_text_as_rich_text() -> None:
    # arrange
    s = "<p>Tobias Fünke has been to <b>Juárez</b>, \nMéxico with his niña &amp; hermosa.</p>"
    # act
    sut = UnitApplicationForTesting(rich_text_as_rich_text=s)
    # assert
    assert_that(sut.rich_text_as_rich_text).described_as(
        "RICH_TEXT should go untouched"
    ).is_equal_to(s)


@pytest.mark.parametrize(
    "field",
    [
        (test_ua_schema.text_as_ascii_text),
        (test_ua_schema.text_multiline_as_ascii_text),
        (test_ua_schema.text_multi_select_as_ascii_text),
        (test_ua_schema.text_multiple_choice_as_ascii_text),
    ],
)
def test_stores_accented_text_as_ascii_text(field: QuickBaseField) -> None:
    # arrange
    sut = UnitApplicationForTesting()
    attr_name = sut.get_attr_from_fid(field.fid)
    # act
    sut.__setattr__(attr_name, "Tobias Fünke has been to Juárez, México with his niña")
    # assert
    assert_that(sut.__getattribute__(attr_name)).is_equal_to(
        "Tobias Funke has been to Juarez, Mexico with his nina"
    )

```

**Comment by tkutcher** [comment]
I like both of these! I think it helps guard against some common programatic assumptions. My only note functionality-wise would be on the unicode side of things we probably want to leave a parameter switch in there to let it be configurable. I could see a case where someone needs to maintain the true characters even if they don't render right in Quickbase.

As far as where to implement it, I think the right spot to put this would be in [`RecordJsonSerializer`](https://github.com/tkutcher/quickbase-client/blob/988b80d40df7e682c80ad8e3b17aadcc19d29b05/src/quickbase_client/orm/serialize.py#L45) just after the linked line, where we _do_ have the field information. That's where we turn the python object (the instance of a `QuickBaseTable`) in to the "shape" needed to send to the API - it then gets converted to true JSON (i.e. dates &rarr; strings) when it gets run through the `QuickBaseJsonEncoder`.  The `QuickBaseTableClient` uses a `RecordJsonSerializer` to create the payload.

Then we can also add the parameter for unicode stuff here.

So something roughly like: 

```py
class RecordJsonSerializer(RecordSerializer):

    def __init__(self, table_cls: Type[QuickBaseTable], normalize_unicode=True):
        self.table_cls = table_cls
        self.normalize_unicode = normalize_unicode

    def serialize(self, record: 'QuickBaseTable') -> Dict:
        o = {}
        for attr, v in record.__dict__.items():
            if attr[0] == '_' or v is None:
                continue
            field_info = record.get_field_info(attr)
            # new logic here:
            if field_info.field_type == Qb.DATE && isinstance(v, datetime):
                v = v.date()
            if self.normalize_unicode && field_info.is_string_like:
                 v = remove_unicode(v)
            o[field_info.fid] = {'value': v}
        return o

```

Tests would then go in the [`TestRecordJsonSerializer`](https://github.com/tkutcher/quickbase-client/blob/988b80d40df7e682c80ad8e3b17aadcc19d29b05/tests/test_orm/test_serialize.py#L29). I very much support arrested development names.

Then we can provide the same `normalize_unicode` argument to the `QuickBaseTableClient` constructor to configure it.

Thoughts? If you wanted to submit a pull request with that I would definitely merge!

**Comment by mklaber** [comment]
For some reason I was thinking the `RecordJsonSerializer` was separate from `QuickBaseJsonEncoder` (and only the latter was used in requests) but yea makes sense. I'll see what I can whip up.

Meanwhile... mind merging and deploying #45 ? that'll let me swap a poetry reference from git to pypi.

**Comment by mklaber** [comment]
How do you feel about `is_string_like` returning `False` for `RICH_TEXT` ?

**Comment by tkutcher** [comment]
`RecordJsonSerializer` is used at a higher level for turning a `QuickBaseTable` instance (record) to JSON payload corresponding to that table. `QuickBaseJsonEncoder` is used by the lower level API classes like `QuickBaseApiClient` (which the higher level `QuickBaseTableClient` wraps) to get dates and whatnot in the payload to the expected Quickbase formats.

Good point on the `is_string_like` property. Yea maybe just leave the logic as is referencing the specific `field_types`. Can always refactor later.

I went ahead and merged #45 to dev. I was thinking to just deploy them both together - I'll do that as soon as it's ready!
